algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/treatment/HasHadTreatmentWithDrugWithCycles.kt

```diff
@@ -0,0 +1,73 @@
+package com.hartwig.actin.algo.evaluation.treatment
+
+import com.hartwig.actin.algo.evaluation.EvaluationFactory
+import com.hartwig.actin.algo.evaluation.EvaluationFunction
+import com.hartwig.actin.algo.evaluation.treatment.MedicationFunctions.createTreatmentHistoryEntriesFromMedications
+import com.hartwig.actin.algo.evaluation.util.Format.concatItemsWithAnd
+import com.hartwig.actin.algo.evaluation.util.Format.concatItemsWithOr
+import com.hartwig.actin.datamodel.PatientRecord
+import com.hartwig.actin.datamodel.algo.Evaluation
+import com.hartwig.actin.datamodel.algo.EvaluationResult
+import com.hartwig.actin.datamodel.clinical.treatment.Drug
+import com.hartwig.actin.datamodel.clinical.treatment.DrugTreatment
+
+class HasHadTreatmentWithDrugWithCycles(private val drugsToFind: Set<Drug>, private val minCycles: Int?) : EvaluationFunction {
+
+    override fun evaluate(record: PatientRecord): Evaluation {
+        val effectiveTreatmentHistory = record.oncologicalHistory + createTreatmentHistoryEntriesFromMedications(record.medications)
+        val namesToMatch = drugsToFind.map { it.name.lowercase() }.toSet()
+        val drugList = concatItemsWithOr(drugsToFind)
+
+        val drugsByEvaluationResult: Map<EvaluationResult, Set<Drug>> = effectiveTreatmentHistory
+            .flatMap { entry ->
+                val result = if (minCycles == null) EvaluationResult.PASS else when (entry.treatmentHistoryDetails?.cycles) {
+                    null -> EvaluationResult.UNDETERMINED
+                    in minCycles..Int.MAX_VALUE -> EvaluationResult.PASS
+                    else -> EvaluationResult.FAIL
+                }
+                entry.allTreatments()
+                    .mapNotNull { it as? DrugTreatment }
+                    .flatMap { it.drugs }
+                    .filter { it.name.lowercase() in namesToMatch }
+                    .map { result to it }
+            }
+            .groupBy({ it.first }, { it.second })
+            .mapValues { it.value.toSet() }
+
+        return when {
+            drugsByEvaluationResult[EvaluationResult.PASS]?.isNotEmpty() == true -> {
+                val matchingDrugs = drugsByEvaluationResult[EvaluationResult.PASS]!!
+                val cyclesString = minCycles?.let { " for at least $minCycles cycles" } ?: ""
+                EvaluationFactory.pass("Has received treatments with ${concatItemsWithAnd(matchingDrugs)}$cyclesString")
+            }
```

Comment:
I always look to avoid using `!!` even though it's reasonable in this case. Here, you could pre-assign all the groups and then check them in the `when` block. With `groupBy`, we can just do a null check since it will never create an empty group:
```suggestion
        val (drugsMatchingCycles, drugsWithUnknownCycles, drugsNotMatchingCycles) =
            listOf(EvaluationResult.PASS, EvaluationResult.UNDETERMINED, EvaluationResult.FAIL).map(drugsByEvaluationResult::get)

        return when {
            drugsMatchingCycles != null -> {
                val cyclesString = minCycles?.let { " for at least $minCycles cycles" } ?: ""
                EvaluationFactory.pass("Has received treatments with ${concatItemsWithAnd(drugsMatchingCycles)}$cyclesString")
            }
```